sockeye/inference.py

```diff
@@ -379,23 +400,26 @@ def load_models(context: mx.context.Context,
                                         vocabulary.  Used when logits/softmax are handled separately.
     :param cache_output_layer_w_b: Models cache weights and biases for logit computation as NumPy arrays (used with
                                    restrict lexicon).
-    :return: List of models, source vocabulary, target vocabulary.
+    :return: List of models, source vocabulary, target vocabulary, source factor vocabularies.
     """
     models, source_vocabs, target_vocabs = [], [], []
     if checkpoints is None:
         checkpoints = [None] * len(model_folders)
     for model_folder, checkpoint in zip(model_folders, checkpoints):
         source_vocabs.append(vocab.vocab_from_json_or_pickle(os.path.join(model_folder, C.VOCAB_SRC_NAME)))
         target_vocabs.append(vocab.vocab_from_json_or_pickle(os.path.join(model_folder, C.VOCAB_TRG_NAME)))
-        model = InferenceModel(model_folder=model_folder,
-                               context=context,
-                               beam_size=beam_size,
-                               batch_size=batch_size,
-                               softmax_temperature=softmax_temperature,
-                               checkpoint=checkpoint,
-                               decoder_return_logit_inputs=decoder_return_logit_inputs,
-                               cache_output_layer_w_b=cache_output_layer_w_b)
-        models.append(model)
+        num_factors = model.SockeyeModel.load_config(os.path.join(model_folder, C.CONFIG_NAME)).num_factors
+        source_factor_vocabs = load_source_factor_vocabs(model_folder, num_factors)
```

Comment:
this is strange: for multiple models it will return the last source_factor_vocabs, whereas for source/target vocabs we are returning the first. There should also be a check that all factor vocabs across models are identical (as in line 424).

Also, you do not seem to need the `num_factors` variable so you can avoid reloading the config from disk. It's available from the model instance.

Reply:
done